kaboxer.1.md
% KABOXER(1)
% Roland Mas
% 2019-2020

# NAME

**kaboxer** - Manage Kali Applications in containers

# SYNOPSIS

**kaboxer** run [**--component** *COMPONENT*] [**--reuse-container**] [**--detach**] [**--prompt-before-exit**] [**--version** *VERSION*] *APP* [*EXECUTABLE*]...

**kaboxer** stop [**--component** *COMPONENT*] [**--prompt-before-exit**] *APP*

**kaboxer** get-meta-file *APP* *FILE*

**kaboxer** get-upstream-version *APP*

**kaboxer** prepare *APP*...

**kaboxer** upgrade *APP*...

**kaboxer** list [**--installed**] [**--available**] [**--upgradeable**] [**--all**] [**--skip-headers**]

**kaboxer** build [**--skip-image-build**] [**--save**] [**--push**] [**--version** *VERSION*] [**--ignore-version**] [*APP*] [*PATH*]

**kaboxer** install [**--tarball**] [**--destdir** *DESTDIR*] [**--prefix** *PREFIX*] [*APP*] [*PATH*]

**kaboxer** clean [*APP*] [*PATH*]

**kaboxer** pull *APP*...

**kaboxer** push [**--version** *VERSION*] *APP*

**kaboxer** save *APP* *FILE*

**kaboxer** load *APP* *FILE*

**kaboxer** purge [**--prune**] *APP*

# DESCRIPTION

**kaboxer** is a framework to build, deploy, install and run images of
containerized applications.  The tool uses subcommands, each with
their own arguments and options.

# GENERAL OPTIONS

**-h**, **--help**
:   Display a friendly help message.

**-v**, **--verbose**
:   Increase verbosity (can be used several times).

# KABOXER RUN

**kaboxer** run [**--component** *COMPONENT*] [**--reuse-container**] [**--detach**] [**--prompt-before-exit**] [**--version** *VERSION*] *APP* [*EXECUTABLE*]...

The run mode is how you actually run a kaboxed application *APP*. In
case several components exist for this app, use
**--component** *COMPONENT* to choose which one to run; if these
components are intended to run in the same container (rather than in
different but communicating containers), use **--reuse-container**. If
the component is meant to run in the background as a daemon, use
**--detach** flag, possibly with the **--prompt-before-exit** flag
(which waits a user confirmation before exiting, so that the user has the
time to read any message displayed or so that applications started in the
after\_run hook are not immediately closed). In case several versions of
the app are installed, the **--version** *VERSION* option can be used to
select which one to actually run. The *EXECUTABLE* parameter can be used
to pass extra arguments to the application.

# KABOXER STOP

**kaboxer** stop [**--component** *COMPONENT*] [**--prompt-before-exit**] *APP*

Stops a running container for application *APP*. Specify
**--component** *COMPONENT* if the app is multi-component to select
which one to stop. Use **--prompt-before-exit** to wait a user
confirmation before exiting.

# KABOXER GET-META-FILE

**kaboxer** get-meta-file *APP* *FILE*

Extracts medadata file *FILE* from the container image for application
*APP*. Mostly used internally by **kaboxer**.

# KABOXER GET-UPSTREAM-VERSION

**kaboxer** get-upstream-version *APP*

Extracts the upstream version of installed application *APP*.

# KABOXER PREPARE

**kaboxer** prepare *APP*...

Runs whatever steps are needed to make the container for application
*APP* ready to run: unpacking an image from a file, pulling from a
registry, and so on. Multiple applications can be prepared at the same time.

# KABOXER UPGRADE

**kaboxer** upgrade *APP*...

Brings the installed image for application *APP* to the latest
version. As for **kaboxer**_prepare_, multiple applications can be
handled at the same time.

# KABOXER LIST

**kaboxer** list [**--installed**] [**--available**] [**--upgradeable**] [**--all**] [**--skip-headers**]

Displays lists of known applications, either a single category
(**--installed** and so on), or all applications (**--all**). The
**--skip-headers** flag hides the column headers to allow for
scriptability.

# KABOXER BUILD

**kaboxer** build [**--skip-image-build**] [**--save**] [**--push**] [**--version** *VERSION*] [**--ignore-version**] [*APP*] [*PATH*]

Builds **kaboxer** images for applications. Unless an application
*APP* is specified, builds all applications found in directory *PATH*
(uses the current directory if unspecified).  With **--save**, saves
the image as a tarball. With **--push**, pushes the image to its
configured registry. With **--version** *VERSION*, passes a version
number to the build process to build an image for a specific
version. With **--ignore-version**, ignores version checks embedded in
the \*.kaboxer.yaml file. With **--skip-image-build**, only builds
the desktop files and does not try to build the container image.

# KABOXER INSTALL

**kaboxer** install [**--tarball**] [**--destdir** *DESTDIR*] [**--prefix** *PREFIX*] [*APP*] [*PATH*]

Installs image and related files (\*.desktop files and icons). With
**--tarball**, uses an image saved as a tarball.  With
**--destdir** *DESTDIR*, use a specific installation directory (can be
used for packaging).
If *APP* is omitted, install all images
corresponding to \*.kaboxer.yaml files found in path *PATH* (current
directory if omitted).

# KABOXER CLEAN

**kaboxer** clean [*APP*] [*PATH*]

Removes files generated as part of **kaboxer** build (tarballs and
generated \*.desktop files). If *APP* is omitted, cleans all apps
corresponding to \*.kaboxer.yaml files found in path *PATH* (current
directory if omitted).

# KABOXER PUSH

**kaboxer** push [**--version** *VERSION*] *APP*

Push an image for application *APP* to its configured registry. If
**--version** *VERSION* is specified, push a specific version.

# KABOXER SAVE

**kaboxer** save *APP* *FILE*

Save the image for application *APP* into a tarball at *FILE*.

# KABOXER LOAD

**kaboxer** load *APP* *FILE*

Loads the image for application *APP* from a tarball at *FILE*.

# KABOXER PURGE

**kaboxer** purge [**--prune**] *APP*

Uninstalls application *APP* completely.


debian/changelog
kaboxer (0.7.1) kali-dev; urgency=medium

  * Fix missed flake8 error
  * Do not run flake8 tests during build

 -- Raphaël Hertzog <raphael@offensive-security.com>  Thu, 10 Dec 2020 22:52:09 +0100

kaboxer (0.7.0) kali-dev; urgency=medium

  * Add support for hook scripts
  * Move exit prompt after after_run hook
  * Replace --prompt-before-detach by --prompt-before-exit
  * Do not rewrite the YAML file when we don't ship a tarball
  * Ensure kaboxer push also updates the "latest" tag when needed
  * Fix most Python coding style issues in the testsuite
  * Document the {before,after}_{run,stop}_script config keys
  * Drop useless dependencies on confget and python3-distutils
  * Make kaboxer/kbxbuilder importable as a library
  * Switch kaboxer into a full-blown Python project
  * Move dh-kaboxer back to debian packaging
  * Move manual pages into a sub-directory
  * Extend the Python metadata
  * Move Python modules to build dependencies
  * Drop unused /etc/kaboxer directory
  * Fix copyright file to mention Offensive Security
  * Add unit tests framework and basic tox configuration
  * More style cleanup to please flake8
  * Enable tests with tox in debian/rules
  * Avoid direct usage of dpkg-parsechangelog
  * Add ${perl:Depends} for dh_kaboxer
  * Provide a longer description in debian/control
  * Switch to debhelper compat level 13
  * Bump Standards-Version to 4.5.1
  * Add logging to help in debugging

 -- Raphaël Hertzog <raphael@offensive-security.com>  Thu, 10 Dec 2020 22:36:16 +0100

kaboxer (0.6.3) kali-dev; urgency=medium

  * Don't fail at startup if we don't have docker access

 -- Raphaël Hertzog <raphael@offensive-security.com>  Thu, 22 Oct 2020 22:29:31 +0200

kaboxer (0.6.2) kali-dev; urgency=medium

  [ Roland Mas ]
  * Added kaboxer(1), kaboxer.yaml(5), kbxbuilder(1),
    kbxbuilder.config.yaml(5) and kbxbuilder.apps.yaml(5) man pages.
  * Add "make clean" to remove generated manpages
  * Use /data rather than /var/lib/kbx-hello inside the container
  * Document run_as_root and publish_ports

  [ Raphaël Hertzog ]
  * Implement DH_KABOXER_BUILD_STRATEGY
  * Add manual pages to the binary package
  * Fix whatis entry in manual pages
  * Define the default value of the path argument at the argument parsing level
  * Implement "kaboxer build --skip-image-build" and use that in dh-kaboxer
  * Add some common gitlab-ci rules to build a docker image for a kaboxer app

 -- Raphaël Hertzog <raphael@offensive-security.com>  Thu, 22 Oct 2020 18:09:54 +0200

kaboxer (0.6.1) kali-dev; urgency=medium

  * Display details of connection exception in debug mode
  * Refactor the code querying the remote docker registry over HTTP
  * Fallback to latest tag when we can't identify a version to retrieve

 -- Raphaël Hertzog <raphael@offensive-security.com>  Fri, 18 Sep 2020 19:04:30 +0200

kaboxer (0.6) kali-dev; urgency=medium

  [ Roland Mas ]
  * dh_kaboxer: decorrelate kaboxer apps from binary packages.
  * Removed debugging message
  * Only clean generated *.desktop files (fixes #29)
  * Don't fetch info about unrelated apps in kaboxer prepare (fixes #30)
  * Fix "kaboxer upgrade foo=1.1"
  * Gracefully handle URL errors in registries
  * Handle API-less registries (registry.hub.docker.com) (fixes #28)
  * Tag image downloaded as 'latest' with its actual version (fixes #31)
  * Remove debugging output
  * Simplify and fix dh_kaboxer injections (fixes #34)
  * Properly handle --prompt-before-detach (fixes #33)
  * Implement kaboxer stop --component (fixes #32)
  * Remove temporary container on error (fixes #35)
  * Fix calculated component name and testsuite
  * Implement kaboxer get-upstream-version (fixes #36)
  * Use proper API to get list of images on a registry (fixes #39)
  * Display packaging revisions in kaboxer list
  * Implement kaboxer list --skip-headers for scriptability (fixes #37)
  * Use the "current" tag rather than "latest" (fixes #38)
  * Fix running of freshly-built image (fixes #43)

  [ Raphaël Hertzog ]
  * Follow best practices for Dockerfiles
  * Use debhelper compat 13 in kbx-hello sample
  * Let dh_auto_* invocations forward supplementary parameters to the kaboxer
    command line
  * Display exception details and URL that we failed to pull
  * Use only the first line of the description in the Comment field
  * Fix pyflakes error reporting usage of unknown function
  * Use --prompt-before-detach instead of --prompt_before_detach
  * Add a missing space in the kaboxer run command line
  * Install extracted icon in kbx-hello-allinone
  * Do not bail out when description or categories are missing
  * Use registry.hub.docker.com as registry to have a realistic sample
  * Fix most Python coding style issues
  * Wrap long lines to 80 characters
  * dh-kaboxer: don't build the tarball by default

 -- Raphaël Hertzog <raphael@offensive-security.com>  Fri, 11 Sep 2020 17:39:42 +0200

kaboxer (0.5.1) unstable; urgency=medium

  * dh_kaboxer: only inject stuff (maintainer scripts, dependencies) in
    relevant packages.

 -- Roland Mas <roland.mas@gnurandal.com>  Tue, 02 Jun 2020 18:05:00 +0200

kaboxer (0.5) kali-dev; urgency=medium

  * Handle versioning and upgrades.

 -- Roland Mas <roland.mas@gnurandal.com>  Fri, 01 May 2020 17:35:12 +0200

kaboxer (0.4) kali-dev; urgency=medium

  * Implement kaboxer as a standalone framework.

 -- Roland Mas <roland.mas@gnurandal.com>  Mon, 23 Mar 2020 15:16:04 +0100

kaboxer (0.3.4) kali-dev; urgency=medium

  * Allow specifying options for Docker.

 -- Roland Mas <roland.mas@gnurandal.com>  Mon, 02 Mar 2020 12:39:18 +0100

kaboxer (0.3.3) kali-dev; urgency=medium

  * Added autopkgtest for kbx-hello-* packages.

 -- Roland Mas <roland.mas@gnurandal.com>  Fri, 14 Feb 2020 14:00:54 +0100

kaboxer (0.3.2) unstable; urgency=medium

  * Expand tildes in mounts to allow persisting the home directory (or
    part thereof) across runs.

 -- Roland Mas <roland.mas@gnurandal.com>  Thu, 13 Feb 2020 17:32:12 +0100


kaboxer.py
#! /usr/bin/python3

import argparse
import yaml
import os
import sys
import re
import grp
import subprocess

import docker
import dockerpty

class Kaboxer:
    def __init__(self):
        self.parser = argparse.ArgumentParser()

        subparsers = self.parser.add_subparsers(title='subcommands', help='action to perform', dest='action', required=True)

        parser_run = subparsers.add_parser('run', help='run containerized app')
        parser_run.add_argument('app')
        parser_run.add_argument('executable', nargs='*')
        parser_run.set_defaults(func=self.run)

        parser_stop = subparsers.add_parser('stop', help='stop running containerized app')
        parser_stop.add_argument('app')
        parser_stop.set_defaults(func=self.stop)

        parser_build = subparsers.add_parser('build', help='build image')
        parser_build.add_argument('app')
        parser_build.set_defaults(func=self.build)

        parser_pull = subparsers.add_parser('pull', help='fetch image')
        parser_pull.add_argument('app')
        parser_pull.set_defaults(func=self.pull)

        parser_purge = subparsers.add_parser('purge', help='purge image')
        parser_purge.add_argument('app')
        parser_purge.set_defaults(func=self.purge)

        try:
            # print ("Getting conn")
            self.docker_conn = docker.from_env()
            # print ("Getting list")
            self.docker_conn.containers.list()
            # print ("List OK")
        except:
            groups = list(map(lambda g: grp.getgrgid(g)[0], os.getgroups()))
            if 'docker' in groups:
                print ("No access to Docker even though you're a member of the docker group, is docker.service running?")
                sys.exit(1)
            else:
                if 'kaboxer' in groups:
                    nc = ['sudo', '-g', 'docker'] + sys.argv
                    # print ("Sudoing with "+str(nc))
                    sys.stdout.flush()
                    sys.stderr.flush()
                    os.execv('/usr/bin/sudo', nc)
                else:
                    print ("No access to Docker, are you a member of group docker or kaboxer?")
                    sys.exit(1)

    def go(self):
        self.args = self.parser.parse_args()
        self.args.func()

    def run(self):
        self.read_config()
        # print(self.config)
        opts = {}
        opts['environment'] = {}
        opts['auto_remove'] = True
        try:
            netname = self.config['networks'][0]
            self.create_network(netname)
            opts['network'] = netname
            extranets = self.config['networks'][1:]
        except KeyError:
            pass
        try:
            ports = {}
            for publish_port in self.config['publish_ports']:
                ports[publish_port] = publish_port
            opts['ports'] = ports
        except KeyError:
            pass
        try:
            mounts = list(map (lambda x: docker.types.Mount(x['target'],x['source'],type='bind'), self.config['mounts']))
            opts['mounts'] = mounts
        except KeyError:
            pass
        run_mode = self.config['run_mode']
        try:
            image = self.config['image']
        except KeyError:
            image = self.args.app

        try:
            opts['command'] = self.config['extra_opts']
        except KeyError:
            pass

        if run_mode == 'cli':
            opts['tty'] = True
            opts['stdin_open'] = True
        elif run_mode == 'gui':
            xsock = '/tmp/.X11-unix'
            xauth_out = os.path.join(os.getenv('HOME'), '.docker.xauth')
            xauth_in = '/root/.docker.xauth'
            f = subprocess.Popen(['xauth', 'nlist', os.getenv('DISPLAY')], stdout=subprocess.PIPE).stdout
            g = subprocess.Popen(['xauth', '-f', xauth_out, 'nmerge', '-'], stdin=subprocess.PIPE).stdin
            for l in f:
                l = str(l,'utf-8')
                l.strip()
                ll = re.sub('^[^ ]*', 'ffff', l) + "\n"
                g.write(bytes(ll, 'utf-8'))
            g.close()
            f.close()
            opts['environment']['DISPLAY'] = os.getenv('DISPLAY')
            opts['environment']['XAUTHORITY'] = xauth_in
            opts['mounts'].append(docker.types.Mount(xauth_in,xauth_out,type='bind'))
            opts['mounts'].append(docker.types.Mount(xsock,xsock,type='bind'))
        elif run_mode == 'server':
            opts['tty'] = True
            opts['name'] = self.args.app
        else:
            print ("Unknown run mode")
            sys.exit(1)
        # print(opts)
        container = self.docker_conn.containers.create(image, *self.args.executable, **opts)
        for e in extranets:
            create_network(e).connect(container)
        dockerpty.start(self.docker_conn.api,container.id)

    def stop(self):
        run_mode = self.config['run_mode']
        try:
            image = self.config['image']
        except KeyError:
            image = self.args.app
        if run_mode == 'server':
            containers = self.docker_conn.containers.list(filter={'name': self.args.app})
            container = containers[0]
            container.stop()
        else:
            print ("Can't stop a non-server component")
            sys.exit(1)

    def build(self):
        self.docker_conn.images.build(path=os.path.join('/usr/src/kaboxer', self.args.app))

    def pull(self):
        self.build()

    def purge(self):
        self.docker_conn.images.remove(self.args.app)

    def read_config(self):
        paths = [
            '.',
            '/etc/kaboxer',
            '/usr/share/kaboxer',
        ]
        for p in paths:
            config_file = os.path.join(p,self.args.app+'.yaml')
            if os.path.isfile(config_file):
                try:
                    self.config = yaml.safe_load(open(config_file))
                    return
                except:
                    print("Error loading config file "+config_file)
                    sys.exit(1)
        print("Missing config file")
        sys.exit(1)

    def create_network(self, netname):
        for n in self.docker_conn.networks.list():
            if n.name == netname:
                return n
        return self.docker_conn.networks.create(name=netname, driver='bridge')

kaboxer = Kaboxer()
kaboxer.go()
